code/iomethods.py

```diff
@@ -518,5 +525,164 @@ def decode_time_from_string(timeString):
+                  of temperature (Kelvin) as an unsigned integer byte, offset by 75 so it will fit in the 0-255 range.
+
+                  For documentation, see http://www.cpc.ncep.noaa.gov/products/global_precip/html/README
+
+        Input:: Path - The path to the MERG binary file
+                Shape - The shape we want the data to be in
+                Offset - The temperatures were scaled to fit into 1-byte by subtracting 75, so the offset is used to
+                         add 75 back to each temperature.
+        Output:: lon - A numPy array containing longitudes from 0 to 360 degrees
+                 lat - A numpy array containing latitudes from 60 to -60 degrees
+
+        Assumption::The binary file was unmodified when downloaded. The shape tuple that is hardcoded in the parameter
+                    will always be the same unless the data changes.
+
+    '''
+    pixel_file = open(path, 'rb')
+    pixel_array = np.fromfile(pixel_file, dtype=np.uint8, count=-1)  # count=-1 means read entire file
+    temperatures = pixel_array.astype(np.float).reshape(shape)
+    temperatures += offset
+
+    lon = np.arange(0.0182, 360., 0.036378335, dtype=np.float)
+    lat = np.arange(59.982, -60., -0.036383683, dtype=np.float)
+
+    pixel_file.close()
+
+    return lon, lat, temperatures
+
+
+# TODO this seems to be unnecessary but I'll leave it for now just in case
+# FORMAT_DEFS = {
+#     "trmm" :
+#         {
+#             "longitude" : "longitude",
+#             "latitude" : "latitude",
+#             "data_variable" : "pcp"
+#         },
+#     "mtsat" :
+#         {
+#             "longitude": "longitude",
+#             "latitude": "latitude",
+#             "data_variable": "albedo"
+#         }
+# }
+
+def read_netCDF_to_array(filename, variable_to_extract, min_lat, max_lat, min_lon, max_lon,
+                         min_t, max_t):
+
+    dataset = netCDF4.Dataset(filename, 'r', format='NETCDF4')
+
+    extracted_variable = dataset.variables[variable_to_extract][:, :, :]
+    lats_list = dataset.variables['latitude'][:]
+    lons_list = dataset.variables['longitude'][:]
+
+    # Reformatting longitude format to [-180, 180]
+    # TODO is this necessary and correctly done?
+    lons_list[lons_list > 180] = lons_list[lons_list > 180] - 360.
+
+    # TODO rearrange variables? is it necessary? Are we even going to have files with more than one timestamp?
+
+    # TODO verify lat/lon is represented in dataset
+
+    # TODO verify time range is represented in dataset.
+    # This might be a problem, some data has no time encoded. Are we supposed to
+    # scan a folder? Assume from filenames?
+
+    dataset.close()
+
+    return ma.masked_array(extracted_variable)
+
+# **********************************************************************************************************************
+
+if __name__ == '__main__':  # Testing for write_np_array_to_ncdf
```

Comment:
Hi @kwhitehall and @caocampb, I have to agree with you Kim, but this is way larger than a unit test.  It should still be in the test folder, but we should flag it as a module test or end-to-end test.  I see what Campell was doing by wrapping the code in the if __name__ statement because the code won't be run if you import iomethods.py but it will run if you run python iomethods.py.

Campbell my recommendation is to cut this entire code block and paste it into a file called iomethods_module_test.py you will need to import iomethods.py and add in a docstring to explain the purpose of this test, and you should be good to go.  One last thing, in the final version you will want to change that hardcoded path on your machine to something generic like '/path/to/MERG/data/' :+1: